<a href="https://colab.research.google.com/github/FrilantikaKusuma/Data-Analysis-of-the-Global-Disaster-Titanic/blob/main/Assignment_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Investigasi sampel data titanic berikut dengan cara :
1. Cek secara head, tail, sample, info lalu observasi apa yang bisa anda peroleh ?
2. Lakukan Statistical Summary dengan mengekstrak informasi yang didapat dari observasi anda ?
3. Cek apakah ada duplikat dan bagaimana handlenya ?
4. Cek apakah ada missing value, berapa persentasenya jika ada, dan bagaimana cara handlenya ?

## Import Libraries

In [ ]:

import pandas as pd


In [ ]:
# import data
df = pd.read_excel('titanic.xlsx')
df.head()

survived                                             name     sex      age
0         1                    Allen, Miss. Elisabeth Walton  female  29.0000
1         1                   Allison, Master. Hudson Trevor    male   0.9167
2         0                     Allison, Miss. Helen Loraine  female   2.0000
3         0             Allison, Mr. Hudson Joshua Creighton    male  30.0000
4         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0000

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# showing top 5 rows of the data
df.tail()

survived                                     name     sex   age
495         1  Mallet, Mrs. Albert (Antoinette Magnin)  female  24.0
496         0        Mangiavacchi, Mr. Serafino Emilio    male   NaN
497         0               Matthews, Mr. William John    male  30.0
498         0                Maybery, Mr. Frank Hubert    male  40.0
499         0                McCrae, Mr. Arthur Gordon    male  32.0

In [ ]:
#sample 5 rows of the data
# the results change everytime the code is executed (random sample)
df.sample(5)

survived                          name     sex   age
120         1  Frauenthal, Mr. Isaac Gerald    male  43.0
260         1     Shutes, Miss. Elizabeth W  female  40.0
15          0           Baumann, Mr. John D    male   NaN
299         0      Uruchurtu, Don. Manuel E    male  40.0
226         0     Pears, Mr. Thomas Clinton    male  29.0

Observations:
1. There has two columns of numericals(survived and age) and 2 columns of categorical (name and sex)
2. Sex column seems to contain two distinct values (female OR male), but will confirm later
3. survived is apparently also binary (0,1)
4. No obvious on the data (column name vs its entries), all seems good

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  500 non-null    int64  
 1   name      500 non-null    object 
 2   sex       500 non-null    object 
 3   age       451 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 15.8+ KB


Observation:
1. Data contains 4 columns with 500 rows
2. Just the AGE column have missing values, they will be handled later
3. All dtypes seem OK (appropriate), given teh corresponding column name

Statistical Summary

In [ ]:
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
# group columns names based on type
# it will make our life easier onwards
categoricals = ['sex']

numericals = ['survived','age']


In [ ]:
# syntax numerical statistical summary
df[numericals].describe()

survived         age
count  500.000000  451.000000
mean     0.540000   35.917775
std      0.498897   14.766454
min      0.000000    0.666700
25%      0.000000   24.000000
50%      1.000000   35.000000
75%      1.000000   47.000000
max      1.000000   80.000000

Observation:


*   Overall, the minimum and maximum values make sense for each column
*   Mean ~50% (Median) in AGE is somewhat symmetrical distribution
*   Survived column is boolean/binary column since the values is 1 or 0, no need to conclude its simmetricity. Only need to check balance level






In [ ]:
# Syntax describe meethod on categorical data
df[categoricals].describe()

sex
count    500
unique     2
top     male
freq     288

Observation:


*   Sex have 2 unique value, 'male' and 'female'
*   The most gender of the passenger in Titanic is male (288 rows) and the rest is female




In [ ]:
categoricals

['sex']

In [ ]:
# showing the precise value counts
# this code is especially useful if we have many categorical columns
for col in categoricals:
    print(f'Column: {col}')
    print(df[col].value_counts())

Column: sex
sex
male      288
female    212
Name: count, dtype: int64


In [ ]:
for col in numericals:
  print(f'====={col}=====')
  print(df[col].value_counts(), '\n')

=====survived=====
survived
1    269
0    230
Name: count, dtype: int64 

=====age=====
age
35.0000    63
24.0000    23
30.0000    20
36.0000    19
18.0000    14
           ..
12.0000     1
0.8333      1
0.6667      1
7.0000      1
3.0000      1
Name: count, Length: 73, dtype: int64 



# Cleaning Data

# 1. Duplicate Handling

In [ ]:
len(df)

500

In [ ]:
len(df.drop_duplicates())

499

In [ ]:
len(df.drop_duplicates()) / len(df)

0.998

In [ ]:
list(df.columns)

['survived', 'name', 'sex', 'age']

In [ ]:
# Langkah 1 : Ambil baris duplikat (termasuk yang asli)
duplicates = df[df.duplicated(keep=False)]


In [ ]:
duplicates

survived                            name     sex   age
104         1  Eustis, Miss. Elizabeth Mussey  female  54.0
349         1  Eustis, Miss. Elizabeth Mussey  female  54.0

In [ ]:
duplicates.groupby(list(df.columns)).size()

survived  name                            sex     age 
1         Eustis, Miss. Elizabeth Mussey  female  54.0    2
dtype: int64

In [ ]:
duplicates.groupby(list(df.columns)).size().reset_index(name='jumlah duplikat')


survived                            name     sex   age  jumlah duplikat
0         1  Eustis, Miss. Elizabeth Mussey  female  54.0                2

In [ ]:
# Handling Drop Duplicate
df = df.drop_duplicates()

In [ ]:
len(df.drop_duplicates()) / len(df)

1.0

Telah dihandling drop duplikat

# 2. Missing Values Handling


# Identiifying Missing Value

In [ ]:
df.isna().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
df.isnull().sum()

survived     0
name         0
sex          0
age         49
dtype: int64

In [ ]:
for column in df.columns:
    print(f"=========={column}==========")
    display(df[column].value_counts())
    print()


==========survived==========


survived
1    269
0    230
Name: count, dtype: int64


==========name==========


name
McCrae, Mr. Arthur Gordon                                1
Allen, Miss. Elisabeth Walton                            1
Allison, Master. Hudson Trevor                           1
Lehmann, Miss. Bertha                                    1
Laroche, Mrs. Joseph (Juliette Marie Louise Lafargue)    1
                                                        ..
Andrews, Mr. Thomas Jr                                   1
Andrews, Miss. Kornelia Theodosia                        1
Anderson, Mr. Harry                                      1
Allison, Mrs. Hudson J C (Bessie Waldo Daniels)          1
Allison, Mr. Hudson Joshua Creighton                     1
Name: count, Length: 499, dtype: int64


==========sex==========


sex
male      288
female    211
Name: count, dtype: int64


==========age==========


age
24.0000    23
30.0000    20
36.0000    19
18.0000    14
42.0000    14
           ..
12.0000     1
0.8333      1
0.6667      1
7.0000      1
3.0000      1
Name: count, Length: 73, dtype: int64



*   There has 269 survivors and the rest is dead (230)
*   The most gender of in Titanic is male (288) and the rest is female (211)
*   The most of the passenger AGE that survived in range at 18-42 years old


In [ ]:
total_rows = len(df)
total_rows

499

In [ ]:
df.columns

Index(['survived', 'name', 'sex', 'age'], dtype='object')

In [ ]:
# percentage version
total_rows = len(df)

# Menghitung dan menampilkan presentase missing values di setiap kolom satu per satu
for column in df.columns:
    missing_count = df[column].isna().sum()
    missing_percentage = (missing_count / total_rows) * 100
    print(f"Column '{column}' has {missing_count} missing values ({missing_percentage:.2f}%)") # .2f means 2 decimal


Column 'survived' has 0 missing values (0.00%)
Column 'name' has 0 missing values (0.00%)
Column 'sex' has 0 missing values (0.00%)
Column 'age' has 49 missing values (9.82%)


The percentage of missing values below 20% so we handle numerically with median, categorical with mode. But the categorical data type does not have missing values, namely sex.


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       450 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB


In [ ]:
df['sex'].dtype

dtype('O')

In [ ]:
df['sex'].mode()[0]

'male'

In [ ]:
df['survived'].median()

1.0

In [ ]:

# Handling missing value for EDA, without splitting
for column in df.columns:
    if df[column].dtype == 'object':
      # jika kolom bertipe object, isi dengan mode
        df[column].fillna(df[column].mode()[0], inplace=True)
    else: # karena tipe data hanya object dan numerik daja tidak ada yang lain, maka menggunakan else berikut:
          # jika kolom bertipe numerik, isi dengan media
        df[column].fillna(df[column].median(), inplace=True)

<ipython-input-41-010dae67b445>:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[column].fillna(df[column].median(), inplace=True)
<ipython-input-41-010dae67b445>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try 

In [ ]:
df.isna().sum()

survived    0
name        0
sex         0
age         0
dtype: int64

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 499 entries, 0 to 499
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   survived  499 non-null    int64  
 1   name      499 non-null    object 
 2   sex       499 non-null    object 
 3   age       499 non-null    float64
dtypes: float64(1), int64(1), object(2)
memory usage: 19.5+ KB


Success removing NULL values!

# Thankyou and see you in the next analysis!